SliceInferer for slice-by-slice infererence

surajpaib opened the pull request:
Fixes #3458 

### Description
Addition of SliceInferer to the monai.inferers for slice-by-slice inference on a 3D volume using a 2D model. 

### Status
**Work in progress**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

surajpaib commented:
Hi @Nic-Ma, this is the first version of incorporating the `SliceInferer` into `monai.inferers`.

What do you think of the implementation? For now, it is the same as what was in the tutorials with better documentation. I'm trying to think of how to make it more readable and concise. Any inputs would be appreciated.

Nic-Ma commented on docs/source/inferers.rst:
```diff
@@ -36,3 +36,9 @@ Inferers
     :special-members: __call__
+
+`SliceInferer`
+~~~~~~~~~~~~~~~~~
```
Please make the `~~~~` length same as `SliceInferer`.

Thanks.

Nic-Ma commented on monai/inferers/inferer.py:
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
+    Args:
+        spatial_dim: Spatial dimension over which the slice-by-slice inference runs on the 3D volume.
+            For example ``0`` could slide over axial slices. ``1`` over coronal slices and ``2`` over sagittal slices.
+        args: other optional args to be passed to the `__init__` of SliceInferer.
```
I think it should be: "passed to the `__init__` of base class SlidingWindowInferer".

Thanks.

Nic-Ma commented on monai/inferers/inferer.py:
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
+        spatial_dim: Spatial dimension over which the slice-by-slice inference runs on the 3D volume.
+            For example ``0`` could slide over axial slices. ``1`` over coronal slices and ``2`` over sagittal slices.
+        args: other optional args to be passed to the `__init__` of SliceInferer.
+        kwargs: other optional keyword args to be passed to `__init__` of SliceInferer.
```
I think it should be: "passed to the `__init__` of base class SlidingWindowInferer".

Thanks.

Nic-Ma commented on monai/inferers/inferer.py:
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
+        kwargs: other optional keyword args to be passed to `__init__` of SliceInferer.
+
+
+    ``args``, ``kwargs`` follow :py:class:`monai.inferer.SlidingWindowInferer`.
```
After changing the `args` and `kwargs`, no need this line anymore.

Thanks.

Nic-Ma commented on monai/inferers/inferer.py:
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
+            args: optional args to be passed to ``network``.
+            kwargs: optional keyword args to be passed to ``network``.
+        """
+        assert self.spatial_dim < 3, "`spatial_dim` can only be `[D, H, W]` with `0, 1, 2` respectively"
```
Please change to use:
```py
if self.spatial_dim < 3:
    raise ValueError("`spatial_dim` can only be `[H, W, D]` with `0, 1, 2` respectively.")
```

Nic-Ma commented on monai/inferers/inferer.py:
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
+        assert self.spatial_dim < 3, "`spatial_dim` can only be `[D, H, W]` with `0, 1, 2` respectively"
+
+        # Check if roi size (eg. 2D roi) and input volume sizes (3D input) mismatch
+        if len(self.roi_size) != len(inputs.shape[2:]):
```
Could you please help add some doc-string to describe the expected `roi_size`?

Thanks.

Nic-Ma commented on monai/inferers/inferer.py:
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
+            else:
+                raise RuntimeError("Currently, only 2D `roi_size` is supported, cannot broadcast to volume. ")
+
+        return super().__call__(inputs, lambda x: self.network_wrapper(network, x))
```
Seems missed to pass the `args` and `kwargs` to `network_wrapper`.

Thanks.

Nic-Ma reviewed (commented):
Thanks for the quick implementation!
Put some minor comments inline.
And please add some **unit tests** to cover the logic.

Thanks.

surajpaib commented:
Hi @Nic-Ma 

Incorporated all the changes you suggested and added a unit test. Let me know if something is missing. 

I also changed the logic a bit to make the purpose of the `SliceInferer` more clear. The old version would fallback and run the `SlidingWindowInferer` if the `roi_size` was not provided according to the `SliceInferer` requirements - this might be confusing to the user, therefore I've changed it to throw a `RuntimeError` in case the `roi_size` is not as expected. Added a note to describe what `roi_size` expects too.

Nic-Ma commented on monai/inferers/inferer.py:
```diff
@@ -222,3 +222,60 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
+            kwargs: optional keyword args to be passed to ``network``.
+        """
+        if self.spatial_dim > 2:
+            raise ValueError("`spatial_dim` can only be `[H, W, D]` with `0, 1, 2` respectively.")
```
I feel maybe the message should be "`spatial_dim` can only be `[0, 1, 2]` with `[H, W, D]` respectively."?

Thanks.

Nic-Ma reviewed (approved):
Hi @surajpaib ,

Really sorry for the late response, I missed this PR last week..
Looks good to me, put minor comments inline.
@wyli Could you please help double confirm it?

Thanks.

Nic-Ma commented:
Hi @surajpaib ,

There is some format error in the code according to CI tests:
```
monai/inferers/inferer.py:262:16: error: Argument 1 to "len" has incompatible type "Union[Sequence[int], int]"; expected "Sized"  [arg-type]
monai/inferers/inferer.py:263:34: error: Argument 1 to "list" has incompatible type "Union[Sequence[int], int]"; expected "Iterable[int]"  [arg-type]
monai/inferers/inferer.py:268:41: error: Argument 2 to "__call__" of "SlidingWindowInferer" has incompatible type "Callable[[Any], Callable[..., Tensor]]"; expected "Callable[..., Tensor]"  [arg-type]
monai/inferers/inferer.py:268:51: error: Incompatible return value type (got "Callable[..., Tensor]", expected "Tensor")  [return-value]
monai/inferers/inferer.py:281:16: error: Incompatible return value type (got "Tensor", expected "Callable[..., Tensor]")  [return-value]
```

Thanks.
